File: finanze/domain/auto_contributions.py
from datetime import date
from enum import Enum
from typing import Optional
from uuid import UUID

from domain.dezimal import Dezimal
from domain.entity import Entity
from domain.fetch_record import DataSource
from pydantic.dataclasses import dataclass


class ContributionFrequency(str, Enum):
    WEEKLY = "WEEKLY"
    BIWEEKLY = "BIWEEKLY"
    MONTHLY = "MONTHLY"
    BIMONTHLY = "BIMONTHLY"
    EVERY_FOUR_MONTHS = "EVERY_FOUR_MONTHS"
    QUARTERLY = "QUARTERLY"
    SEMIANNUAL = "SEMIANNUAL"
    YEARLY = "YEARLY"


class ContributionTargetType(str, Enum):
    STOCK_ETF = "STOCK_ETF"
    FUND = "FUND"
    FUND_PORTFOLIO = "FUND_PORTFOLIO"


class ContributionTargetSubtype(str, Enum):
    STOCK = "STOCK"
    ETF = "ETF"
    MUTUAL_FUND = "MUTUAL_FUND"
    PRIVATE_EQUITY = "PRIVATE_EQUITY"
    PENSION_FUND = "PENSION_FUND"


@dataclass
class PeriodicContribution:
    id: UUID
    alias: Optional[str]
    target: str
    target_name: str
    target_type: ContributionTargetType
    amount: Dezimal
    currency: str
    since: date
    until: Optional[date]
    frequency: ContributionFrequency
    active: bool
    source: DataSource
    next_date: Optional[date] = None
    target_subtype: Optional[ContributionTargetSubtype] = None
    entity: Optional[Entity] = None


@dataclass
class AutoContributions:
    periodic: list[PeriodicContribution]


@dataclass
class EntityContributions:
    contributions: dict[str, AutoContributions]


@dataclass
class ContributionQueryRequest:
    entities: Optional[list[UUID]] = None
    excluded_entities: Optional[list[UUID]] = None
    real: Optional[bool] = None


@dataclass
class ManualPeriodicContribution:
    entity_id: UUID
    name: str
    target: str
    target_name: Optional[str]
    target_type: ContributionTargetType
    target_subtype: Optional[ContributionTargetSubtype]
    amount: Dezimal
    currency: str
    since: date
    until: Optional[date]
    frequency: ContributionFrequency

File: mobile/src/domain/core/autoContributions.ts
import { Entity } from "./entity"
import { DataSource } from "./fetchRecord"
import { Dezimal } from "@/domain/dezimal"

export enum ContributionFrequency {
  WEEKLY = "WEEKLY",
  BIWEEKLY = "BIWEEKLY",
  MONTHLY = "MONTHLY",
  BIMONTHLY = "BIMONTHLY",
  EVERY_FOUR_MONTHS = "EVERY_FOUR_MONTHS",
  QUARTERLY = "QUARTERLY",
  SEMIANNUAL = "SEMIANNUAL",
  YEARLY = "YEARLY",
}

export enum ContributionTargetType {
  STOCK_ETF = "STOCK_ETF",
  FUND = "FUND",
  FUND_PORTFOLIO = "FUND_PORTFOLIO",
  CRYPTO = "CRYPTO",
}

export enum ContributionTargetSubtype {
  STOCK = "STOCK",
  ETF = "ETF",
  MUTUAL_FUND = "MUTUAL_FUND",
  PRIVATE_EQUITY = "PRIVATE_EQUITY",
  PENSION_FUND = "PENSION_FUND",
}

export interface PeriodicContribution {
  id: string
  alias: string | null
  target: string
  targetName: string
  targetType: ContributionTargetType
  amount: Dezimal
  currency: string
  since: string
  until: string | null
  frequency: ContributionFrequency
  active: boolean
  source: DataSource
  nextDate?: string | null
  targetSubtype?: ContributionTargetSubtype | null
  entity?: Entity | null
}

export interface AutoContributions {
  periodic: PeriodicContribution[]
}

export interface EntityContributions {
  contributions: Record<string, AutoContributions>
}

export interface ContributionQueryRequest {
  entities?: string[] | null
  excludedEntities?: string[] | null
  real?: boolean | null
}

export interface ManualPeriodicContribution {
  entityId: string
  name: string
  target: string
  targetName: string | null
  targetType: ContributionTargetType
  targetSubtype: ContributionTargetSubtype | null
  amount: Dezimal
  currency: string
  since: string
  until: string | null
  frequency: ContributionFrequency
}
